pkg/tfbridge/walk.go

```diff
@@ -168,3 +265,347 @@ func LookupSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *Schema
+		return schemaInfo
+	}
+
+	switch first := schemaPath[0].(type) {
+	case walk.ElementStep:
+		return getSchemaInfoPath(schemaPath[1:], ensureField(&schemaInfo.Elem))
+	case walk.GetAttrStep:
+		return getSchemaInfoPath(schemaPath[1:], ensureMapKey(&schemaInfo.Fields, first.Name))
+	default:
+		contract.Failf("Unable to walk SchemaPath segment of type %T", schemaPath[0])
+		return nil
+	}
+}
+
+type VisitInfo struct {
+	Root VisitRoot
+
+	shimSchema    shim.Schema
+	getSchemaInfo func() *SchemaInfo
+
+	getShimSchemaMap func() shim.SchemaMap
+	getSchemaInfoMap func() map[string]*SchemaInfo
+}
+
+type VisitResult struct {
+	// If the visitor has done something.
+	//
+	// Visits will only be replayed during runtime startup if `HasEffect: true`.
+	HasEffect bool
+}
+
+func (v *VisitInfo) ShimSchema() shim.Schema                        { return v.shimSchema }
+func (v *VisitInfo) SchemaInfo() *SchemaInfo                        { return v.getSchemaInfo() }
+func (v *VisitInfo) EnclosingSchemaInfoMap() map[string]*SchemaInfo { return v.getSchemaInfoMap() }
+func (v *VisitInfo) EnclosingSchemaShimMap() shim.SchemaMap         { return v.getShimSchemaMap() }
+
+type VisitRoot interface {
+	PulumiToken() tokens.Type
+
+	isVisitRoot()
+}
+
+// This ensures that we can peform casts from VisitRoot to T for each valid root. Changing
+// this from T to *T is a breaking change.
+var (
+	_ VisitRoot = VisitResourceRoot{}
+	_ VisitRoot = VisitDataSourceRoot{}
+	_ VisitRoot = VisitConfigRoot{}
+)
+
+type VisitResourceRoot struct {
+	token tokens.Type
+
+	Info *ResourceInfo
+
+	TfToken string
+}
+
+func (r VisitResourceRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitResourceRoot) isVisitRoot()             {}
+
+type VisitDataSourceRoot struct {
+	token tokens.Type
+
+	Info *DataSourceInfo
+
+	TfToken string
+}
+
+func (r VisitDataSourceRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitDataSourceRoot) isVisitRoot()             {}
+
+type VisitConfigRoot struct{ token tokens.Type }
+
+func (r VisitConfigRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitConfigRoot) isVisitRoot()             {}
+
+type Visitor = func(VisitInfo, SchemaPath) (VisitResult, error)
+
+func MustTraverseSchemas(prov *ProviderInfo, traversalID string, visitor Visitor) {
+	err := TraverseSchemas(prov, traversalID, visitor)
+	contract.AssertNoErrorf(err, "failed to traverse provider schemas")
+}
+
+// Traverse all schemas in a provider.
+//
+// traversalId must be unique for each traversal within a `make tfgen`, but the same
+// between `make tfgen` and the provider runtime. A simple descriptive string like `"apply
+// labels"` works great here.
+func TraverseSchemas(prov *ProviderInfo, traversalID string, visitor Visitor) error {
+	prov.MetadataInfo.assertValid() // We must have metadata info for an efficient (sparse) traversal.
+
+	var errs []error
+
+	const key = "traverse-effects"
+	forEffect := currentRuntimeStage == runningProviderStage
```

Comment:
This should be a parameter and not relying on the auto detection, to make the code more testable. Already been running into some trouble with this hidden auto-detector, i.e. benchmarks in pulumi-aws cannot be made to measure the runtime overhead because the auto-detector things that bencharking is happening at the runtime phase.